clippy_lints/src/suspicious_xor.rs

```diff
@@ -0,0 +1,90 @@
+use clippy_utils::{numeric_literal::NumericLiteral, source::snippet};
+use rustc_ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir::{BinOpKind, Expr, ExprKind, Lit};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Warns for a Bitwise XOR operator being probably confused as a powering. It will not trigger if any of the numbers are not in decimal.
+    /// ### Why is this bad?
+    ///	It's most probably a typo and may lead to unexpected behaviours.
+    /// ### Example
+    /// ```rust
+    /// let x = 3_i32 ^ 4_i32;
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let x = 3_i32.pow(4);
+    /// ```
+    #[clippy::version = "1.65.0"]
+    pub SUSPICIOUS_XOR,
+    suspicious,
+    "XOR (`^`) operator possibly used as exponentiation operator"
+}
+declare_lint_pass!(ConfusingXorAndPow => [SUSPICIOUS_XOR]);
+
+impl LateLintPass<'_> for ConfusingXorAndPow {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        if_chain! {
+            if let ExprKind::Binary(op, left, right) = &expr.kind;
+            if op.node == BinOpKind::BitXor;
+            if let ExprKind::Lit(litr) = &right.kind;
+            if let ExprKind::Lit(litl) = &left.kind;
+            if let snip_left = snippet(cx, litl.span, "..");
+            if let snip_right = snippet(cx, litr.span, "..");
+            if get_numlit(litr, &snip_right)
+            .zip(get_numlit(litl, &snip_left))
+            .map_or(false, |(a,b)| a.is_decimal() && b.is_decimal());
+            if let left_val = unwrap_lit_to_dec(left).unwrap_or(0);
+            if let right_val = unwrap_lit_to_dec(right).unwrap_or(0);
+            if let suffix = get_numlit(litr, &snip_right).unwrap().suffix.unwrap_or("");
+                then {
+
+                            if left_val == 2 &&
+                             (right_val == 8 ||
+                            right_val == 16 ||
+                            right_val == 32 ||
+                            right_val == 64 || right_val == 128)
+                            {
+                                clippy_utils::diagnostics::span_lint_and_sugg(
+                                    cx,
+                                    SUSPICIOUS_XOR,
+                                    expr.span,
+                                    "it appears that you are trying to get the maximum value of an integer, but '^' is not exponentiation operator",
```

Comment:
Why is this pointing the max value out? 

It would be better to detect the case where `2^x` and `10^x` like this:
https://github.com/rust-lang/rust-clippy/issues/4205#issuecomment-502075573

Reply:
just an idea I got from #4541 and thought it was killing two birds with one stone